#! /bin/bash

# This file is to extract genotypes based on annotation information
# cd /home/wbi1/one_sided_SKAT/NFBC66_2017_09_18/exon
# bsub -R 'rusage[mem=500]' -J wbi1[1-22] -q priority -P 1234 './Genotypes_1.sh $LSB_JOBINDEX'

if (($#==0))
then
echo "/home/wbi1/one_sided_SKAT/NFBC66_2017_09_18/exon/Genotypes_1.sh CHR"
exit
fi

CHR=$1      # chromosome number, from 1 to 22
# vcf file as input, need complete directory
VCF=/home/wbi1/database/Gkang-dbGaP/phs000276/RootStudyConsent_phs000276.NHLBI_NFBC66.v2.p1.c1.GRU/Data_Process/step_5_QC/chr${CHR}.maf.0.05.R2.0.3.dose.vcf      

# set working directory
DIR=/home/wbi1/one_sided_SKAT/NFBC66_2017_09_18/exon/Genotypes

PREFIX=chr${CHR}   # prefix for all output
# annotation file generated by ANNOVAR; suffix should be ****_multianno.text
ANNO=/home/wbi1/database/Gkang-dbGaP/phs000276/RootStudyConsent_phs000276.NHLBI_NFBC66.v2.p1.c1.GRU/Data_Process/step_6_ANNO/chr${CHR}.maf.0.05.R2.0.3.dose.anno.hg19_multianno.txt

SUBJ=/home/wbi1/one_sided_SKAT/NFBC66_2017_09_18/pheno2.txt

#####################################
## The below is the extract process
#####################################

module load vcftools/0.1.15

mkdir ${DIR} -p
cd ${DIR}

# Use ANNO file to select variants and use PED file to select subjects 
cat ${ANNO} | awk -F "\t" '{if($6~/\yexonic/)print $0}' > ${PREFIX}.anno
cat ${PREFIX}.anno | awk -F "\t" '{print $1"\t"$14}' > ${PREFIX}.anno.pos 

# Use vcftools to generate numeric matrix for genotype
vcftools --vcf ${VCF} --keep ${SUBJ} --out ${PREFIX} --012 --positions ${PREFIX}.anno.pos
vcftools --vcf ${VCF} --keep ${SUBJ} --out ${PREFIX} --freq --positions ${PREFIX}.anno.pos




