
## Pipeline of data analysis

### Input files: 
- VCF file of genotype
- Z-score file generated by ImpG software
- Phenotype file

### Process:
- USE ANNOVAR (http://annovar.openbioinformatics.org/en/latest/) to annotate all alleles
- Generate a mapping file with SNP_id and region_id based on annotation file
- Extract alleles in mapping file to generate FRQ file, raw genotype file, REF and ALT alleles must be specified

### Useful code:
- step-5-QC_rsq_0.3.sh remove imputated variants whose R-sq less than 0.3 and maf greater than 0.05
- Genotypes_1.sh extract genotypes based on annotation information
- Genotypes_R.R generate R objects for further analysis
- main.R final step for analysis


### vcf2R.sh:
- Example code:
```
vcf_file=/home/wbi1/one_sided_SKAT/Update_NFBC66_2018_07_10/chr21.maf.0.05.R2.0.3.dose.vcf
anno_file=/home/wbi1/one_sided_SKAT/Update_NFBC66_2018_07_10/chr21.maf.0.05.R2.0.3.dose.anno.hg19_multianno.txt
output_dir=/home/wbi1/one_sided_SKAT/Update_NFBC66_2018_07_10/Results
output_prefix=chr21.maf.0.05.R2.0.3.dose
vcf2map.sh vcf_file anno_file output_dir output_prefix
```
